File: .claude/hooks/quality_gate.py
#!/usr/bin/env python3
"""
quality_gate (Claude Code hook) — PostToolUse formatter/linter check.

Python port of .kilo/hooks/post-tool-use/quality-gate.js. Stdlib-only for
Windows/macOS/Linux portability (no bash/jq/node required).

Wired via .claude/settings.json:
    "PostToolUse": [ { "matcher": "Edit|Write|MultiEdit",
        "hooks": [ { "type": "command",
            "command": "python .claude/hooks/quality_gate.py" } ] } ]

Behavior:
  - Reads the PostToolUse JSON from stdin: {tool_input:{file_path}, ...}.
  - Runs a NON-BLOCKING format check on the edited file, matched by extension:
      .py                                   -> ruff format --check
      .ts/.tsx/.js/.jsx/.json/.md/.css/...  -> prettier or biome (if configured)
      .go                                   -> gofmt -l
  - On check failure, prints a fix hint to stderr; ALWAYS exits 0 so the edit
    is never blocked (advisory only, mirrors the Kilo hook).
  - Opt out entirely with QUALITY_GATE_DISABLE=1.
  - Missing tools / parse errors / any exception -> silent exit 0.
"""

from __future__ import annotations

import json
import os
import subprocess
import sys
from pathlib import Path

PRETTIER_EXTS = {
    ".ts", ".tsx", ".js", ".jsx", ".mjs", ".cjs",
    ".json", ".jsonc", ".md", ".mdx", ".css", ".scss", ".html", ".yaml", ".yml",
}
_ROOT_MARKERS = (
    "package.json", "pyproject.toml", "go.mod", "Cargo.toml",
    "biome.json", "biome.jsonc", ".prettierrc", ".git",
)


def _run(cmd: list[str], cwd: Path | None = None) -> subprocess.CompletedProcess[str] | None:
    """Run a command, return CompletedProcess or None if the tool is missing."""
    try:
        return subprocess.run(
            cmd,
            cwd=str(cwd) if cwd else None,
            capture_output=True,
            text=True,
            timeout=15,
        )
    except (FileNotFoundError, subprocess.TimeoutExpired, OSError):
        return None


def _find_project_root(start: Path) -> Path:
    cur = start.resolve()
    for parent in (cur, *cur.parents):
        if any((parent / m).exists() for m in _ROOT_MARKERS):
            return parent
    return Path.cwd()


def _detect_js_formatter(root: Path) -> str | None:
    if (root / "biome.json").exists() or (root / "biome.jsonc").exists():
        return "biome"
    for name in (".prettierrc", ".prettierrc.json", ".prettierrc.js",
                 ".prettierrc.yml", ".prettierrc.yaml", "prettier.config.js"):
        if (root / name).exists():
            return "prettier"
    pkg = root / "package.json"
    if pkg.exists():
        try:
            if "prettier" in json.loads(pkg.read_text(encoding="utf-8")):
                return "prettier"
        except (ValueError, OSError):
            pass
    return None


def _warn(msg: str) -> None:
    sys.stderr.write(f"[QualityGate] {msg}\n")


def check(file_path: str) -> None:
    if not file_path:
        return
    path = Path(file_path)
    if not path.exists() or not path.is_file():
        return

    ext = path.suffix.lower()
    root = _find_project_root(path.parent)

    if ext == ".py":
        res = _run(["ruff", "format", "--check", str(path)], cwd=root)
        if res is not None and res.returncode != 0:
            _warn(f"Ruff format check failed for {path.name}. Run: ruff format \"{path}\"")
        return

    if ext in PRETTIER_EXTS:
        formatter = _detect_js_formatter(root)
        if formatter == "prettier":
            res = _run(["npx", "prettier", "--check", str(path)], cwd=root)
            if res is not None and res.returncode != 0:
                _warn(f"Prettier check failed for {path.name}. Run: npx prettier --write \"{path}\"")
        elif formatter == "biome":
            res = _run(["npx", "@biomejs/biome", "check", str(path)], cwd=root)
            if res is not None and res.returncode != 0:
                _warn(f"Biome check failed for {path.name}")
        return

    if ext == ".go":
        res = _run(["gofmt", "-l", str(path)], cwd=root)
        if res is not None and res.stdout.strip():
            _warn(f"gofmt check failed for {path.name}. Run: gofmt -w \"{path}\"")
        return


def main() -> int:
    if os.environ.get("QUALITY_GATE_DISABLE", "").lower() in ("1", "true"):
        return 0
    try:
        payload = json.load(sys.stdin)
    except (json.JSONDecodeError, ValueError):
        return 0
    try:
        tool_input = payload.get("tool_input", {}) or {}
        check(tool_input.get("file_path", "") or "")
    except Exception:  # noqa: BLE001 — advisory hook must never crash the tool call
        pass
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: .kilo/hooks/post-tool-use/quality-gate.js
#!/usr/bin/env node
/**
 * quality-gate.js — PostToolUse hook for Solo-Code-Harness
 *
 * Runs lightweight formatter/linter checks after file edits.
 * Supports: Prettier, Biome, Ruff (Python), gofmt (Go).
 * Ported from ECC's quality-gate.js.
 *
 * Exit codes:
 *   0 = PASS or SKIP
 *   1 = WARNING (non-blocking)
 */

'use strict';

const fs = require('fs');
const path = require('path');
const { spawnSync } = require('child_process');

const MAX_STDIN = 1024 * 1024;

function exec(command, args, cwd = process.cwd()) {
  return spawnSync(command, args, {
    cwd,
    encoding: 'utf8',
    env: process.env,
    timeout: 15000,
    windowsHide: true,
  });
}

function log(msg) {
  process.stderr.write(`${msg}\n`);
}

/**
 * Find project root by looking for common marker files.
 */
function findProjectRoot(startDir) {
  let dir = path.resolve(startDir || process.cwd());
  const markers = ['package.json', 'pyproject.toml', 'go.mod', 'Cargo.toml', 'biome.json', '.prettierrc', '.git'];
  const root = path.parse(dir).root;

  while (dir !== root) {
    if (markers.some(m => fs.existsSync(path.join(dir, m)))) {
      return dir;
    }
    dir = path.dirname(dir);
  }
  return process.cwd();
}

/**
 * Check which formatter is configured in the project.
 */
function detectFormatter(projectRoot) {
  if (fs.existsSync(path.join(projectRoot, 'biome.json'))) return 'biome';
  if (fs.existsSync(path.join(projectRoot, 'biome.jsonc'))) return 'biome';
  if (fs.existsSync(path.join(projectRoot, '.prettierrc'))) return 'prettier';
  if (fs.existsSync(path.join(projectRoot, 'prettier.config.js'))) return 'prettier';
  if (fs.existsSync(path.join(projectRoot, 'package.json'))) {
    try {
      const pkg = JSON.parse(fs.readFileSync(path.join(projectRoot, 'package.json'), 'utf8'));
      if (pkg.prettier) return 'prettier';
    } catch {}
  }
  return null;
}

function runQualityGate(filePath) {
  if (!filePath || !fs.existsSync(filePath)) return;

  filePath = path.resolve(filePath);
  const ext = path.extname(filePath).toLowerCase();
  const projectRoot = findProjectRoot(path.dirname(filePath));
  const formatter = detectFormatter(projectRoot);

  // JS/TS — Prettier or Biome
  if (['.ts', '.tsx', '.js', '.jsx', '.json', '.md'].includes(ext)) {
    if (formatter === 'prettier') {
      const result = exec('npx', ['prettier', '--check', filePath], projectRoot);
      if (result.status !== 0) {
        log(`[QualityGate] Prettier check failed for ${path.relative(process.cwd(), filePath)}. Run: npx prettier --write "${filePath}"`);
      }
    } else if (formatter === 'biome') {
      const result = exec('npx', ['@biomejs/biome', 'check', filePath], projectRoot);
      if (result.status !== 0) {
        log(`[QualityGate] Biome check failed for ${path.relative(process.cwd(), filePath)}`);
      }
    }
  }

  // Python — Ruff
  if (ext === '.py') {
    const result = exec('ruff', ['format', '--check', filePath]);
    if (result.status !== 0) {
      log(`[QualityGate] Ruff check failed for ${path.relative(process.cwd(), filePath)}. Run: ruff format "${filePath}"`);
    }
  }

  // Go — gofmt
  if (ext === '.go') {
    const result = exec('gofmt', ['-l', filePath]);
    if (result.stdout && result.stdout.trim()) {
      log(`[QualityGate] gofmt check failed for ${path.relative(process.cwd(), filePath)}. Run: gofmt -w "${filePath}"`);
    }
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
let raw = '';
let truncated = false;

process.stdin.setEncoding('utf8');
process.stdin.resume();
process.stdin.on('data', chunk => {
  if (raw.length < MAX_STDIN) {
    const remaining = MAX_STDIN - raw.length;
    raw += chunk.substring(0, remaining);
    if (chunk.length > remaining) truncated = true;
  } else {
    truncated = true;
  }
});

process.stdin.on('end', () => {
  process.stdout.write(raw);

  if (truncated) {
    process.exit(0);
  }

  // Disable quality gate if opted out
  if (String(process.env.QUALITY_GATE_DISABLE || '').toLowerCase() === '1') {
    process.exit(0);
  }

  try {
    const input = JSON.parse(raw);
    const filePath = String(input?.tool_input?.file_path || '');
    runQualityGate(filePath);
  } catch {
    // Ignore parse errors
  }

  process.exit(0);
});
